### ISSUE TITLE
Build my own Oracle Prometheus 

### ISSUE THREAD

**Issue by Seth1u** [juev/prometheus_oracle#1]
Hello Juev!

This is a great job creating an Oracle export. I would like to continue your work on my test environment. I want to compile an exporter binary that will work for different versions of Oracle like 12.1, 12.2 and 19.

Do you have an end-to-end build and install procedure for exporting Oracle to a Linux server (Redhat for example)?
I’d like to know if you have any Grafana dashboards that go with it.

Thank you very much

**Comment by juev** [comment]
Hello @Seth1u! 

This repository contains build system based on Makefile. You can run `make build`, or `make linux` for example. And this command should prepare binary file or docker container.

By default each commit to this repository running docker build and publish:

https://hub.docker.com/repository/docker/juev/prometheus_oracle

And you can use docker container without build:

```
docker run --rm -v /Users/username/prometheus_oracle/config.yaml:/config.yaml -v /Users/username/prometheus_oracle/log.json:/log.json --name prometheus_oracle -p 9101:9101 juev/prometheus_oracle
```

You should just prepare your config file.

This exporter prepared data for prometheus. And yes, you can use it with grafana. I dont have examples for you, because its your data.

As I can see on [page](https://www.oracle.com/database/technologies/faq-instant-client.html):

> An Instant Client based application can interoperate with any database version that the full Oracle Client interoperates with. For example, Instant Client 12.2 allows connection to 11.2, 12.1 and 12.2 databases. Details are in the interoperability matrix in Support Doc ID 207303.1 Note that Instant Client includes a number of different Oracle tools and technologies, some of which may have more restrictive requirements.

You can use this exporter with all previous version Oracle.

**Comment by Seth1u** [comment]
Thank you for your response!
I will test it now. And if i need to connect to my database with this exporter without listener (TNSNAME), how can I change  exporter main.go source code for connect to database with another user like:
sqlplus user/password

**Comment by juev** [comment]
@Seth1u 
We have [config.yaml.template](https://github.com/juev/prometheus_oracle/blob/master/config.yaml.template) file. We should rename it to config.yaml, for example. After that we should fill it.

```  
- host: 'dummy'
    user: dummy
    port: 1522
    password: 'password'
    database: dummy
    maxIdleConns: 10
    maxOpenConns: 10
    queries:
      - sql: "select numbers1 from dummy"
        name: value1
        interval: 1
```

`host` -- your database ip or domain name.
`user` -- your username
`port` -- port for yout database, by default oracle works on port 1522
`password` -- your password
`database` -- your database name on `host`
`maxIdleConns` and `maxOpenConns` -- settings for connection pool

`queries` -- query lists, when we shoud provide:
`sql` -- same query
`name` -- name for this query
`interval` -- number minutes, time between starts next query.

You can use this configuration file when you run docker container.

**Comment by Seth1u** [comment]
@juev 
i'm not on docker environment, i will use this exporter on the physical server that lodge all databases.

So, i don't need to connect to database using oracle default port.
Otherwise it will be a good idea to use it with docker container or on the server itself.
Thanks.

**Comment by juev** [comment]
@Seth1u I need a little more time for research sqlplus command line options. After that I will return with answer for you.

In this time prometheus_oracle used dsn:
```
user/password@host:port/db
```

Makefile contains all what ou need for build prometheus_oracle on linux.

Its will install instantclient and instaclient sdk to you computer. Will setup `LD_LIBRARY_PATH` and `PKG_CONFIG_PATH` variables. After that `go build` it will work.

**Comment by Seth1u** [comment]
When i'm try command make linux i have got some errors :
# github.com/mattn/go-oci8
In file included from /root/go/pkg/mod/github.com/mattn/go-oci8@v0.0.4/c_helpers.go:3:0:
./oci8.go.h:1:17: fatal error: oci.h: No such file or directory
 #include <oci.h>

**Comment by Seth1u** [comment]
Thanks you @juev for your availability!
Good job

**Comment by juev** [comment]
@Seth1u this error meens:

1. you didnt install instantclient sdk. For building you should install 2 package: instantclient basic and instantclient sdk.
2. your setup is broken. You need file oci8.pc with filling all paths to instantclient files. Then you need `PKG_CONFIG_PATH` variable, what setuped to path with your oci8.pc file. And you need `LD_LIBRARY_PATH` variable, what setuped to path with instantclient libs path.

You can download instantclient zip packages from [oracle](https://www.oracle.com/database/technologies/instant-client/downloads.html) site (basic and sdk). Unzip it in one folder. And use this folder for building.

**Comment by juev** [comment]
@Seth1u I published release [0.1.0](https://github.com/juev/prometheus_oracle/releases/tag/0.1.0), you can try to use binary file for linux or darwin. But you need instantclient basic and configured variable `LD_LIBRARY_PATH` for running these binaries.

**Comment by Seth1u** [comment]
@juev  I installed your binary file and  i use some linux sevice like this:

[Unit]
Description=Oracle Prometheus Exporter Juev 

[Service]

Environment=ORACLE_HOME=/oracle/product/12.1.0.2
Environment=LD_LIBRARY_PATH=/oracle/product/12.1.0.2/lib
Environment=NLS_LANG=AMERICAN_AMERICA.UTF8

User=root
Restart=on-failure
ExecStart=/usr/local/bin/prometheus_oracle_linux -configfile=/etc/oracle_exporterconfig/oracle_exporter.conf

[Install]
WantedBy=multi-user.target

Is it right configuration?

And i  have configuration on my oracle_exporter.conf like this 

connections:
 - connection: user/password@dummy
   database: dummy
   instance: dummy
   alertlog:
    - file: /oracle/diag/rdbms/dummy/dummy/trace/alert_dummy.log
      ignoreora:
       - ORA-00001
       - ORA-01033
       - ORA-01041
       - ORA-01089
       - ORA-01555
       - ORA-28
       - ORA-235
       - ORA-609
       - ORA-3136
       - ORA-01089
   queries:
    - sql: "select 1 from dual"
      name: sample1

Thanks u

**Comment by Seth1u** [comment]
@juev  I run make
andd output is: 
-rw-r--r-- 1 root root  52M Nov 25 15:52 oracle-instantclient19.5-basic-19.5.0.0.0-1.x86_64.rpm
-rw-r--r-- 1 root root 599K Nov 25 15:52 oracle-instantclient19.5-devel-19.5.0.0.0-1.x86_64.rpm
-rwxr-xr-x 1 root root 2.0K Feb 11 08:12 README.md
-rwxr-xr-x 1 root root  223 Feb 11 08:12 oci8.pc.template
-rwxr-xr-x 1 root root 1.1K Feb 11 08:12 LICENSE
-rwxr-xr-x 1 root root  21K Feb 11 08:12 go.sum
-rwxr-xr-x 1 root root  320 Feb 11 08:12 go.mod
-rwxr-xr-x 1 root root 1.5K Feb 11 08:12 Dockerfile
-rwxr-xr-x 1 root root  604 Feb 11 08:12 config.yaml.template
-rwxr-xr-x 1 root root 2.6K Feb 13 18:27 Makefile.bkp
-rwxr-xr-x 1 root root 5.4K Feb 13 18:34 main.go
-rwxr-xr-x 1 root root 2.6K Feb 14 16:15 Makefile
 

And then i run: 

rpm2cpio oracle-instantclient19.5-basic-19.5.0.0.0-1.x86_64.rpm | cpio -idmv
and  rpm2cpio oracle-instantclient19.5-devel-19.5.0.0.0-1.x86_64.rpm | cpio -idmv

-rw-r--r-- 1 root root  52M Nov 25 15:52 oracle-instantclient19.5-basic-19.5.0.0.0-1.x86_64.rpm
-rw-r--r-- 1 root root 599K Nov 25 15:52 oracle-instantclient19.5-devel-19.5.0.0.0-1.x86_64.rpm
-rwxr-xr-x 1 root root 2.0K Feb 11 08:12 README.md
-rwxr-xr-x 1 root root  223 Feb 11 08:12 oci8.pc.template
-rwxr-xr-x 1 root root 1.1K Feb 11 08:12 LICENSE
-rwxr-xr-x 1 root root  21K Feb 11 08:12 go.sum
-rwxr-xr-x 1 root root  320 Feb 11 08:12 go.mod
-rwxr-xr-x 1 root root 1.5K Feb 11 08:12 Dockerfile
-rwxr-xr-x 1 root root  604 Feb 11 08:12 config.yaml.template
-rwxr-xr-x 1 root root 5.4K Feb 13 18:34 main.go
drwxr-xr-x 1 root root  710 Feb 14 10:39 instantclient_19_5
drwxr-xr-x 1 root root   24 Feb 14 10:49 etc
drwxr-xr-x 1 root root   30 Feb 14 10:50 usr
-rwxr-xr-x 1 root root 2.6K Feb 14 16:15 Makefile

I changed some lines in Makefile and the new Makefile look like this:

VERSION        ?= 0.1.0
ORACLE_VERSION ?= 19.5
LDFLAGS        := -X main.Version=$(VERSION)
GOFLAGS        := -ldflags "$(LDFLAGS) -s -w"
ARCH           ?= $(shell uname -m)
GOARCH         ?= $(subst x86_64,amd64,$(patsubst i%86,386,$(ARCH)))
RPM_VERSION    ?= $(ORACLE_VERSION).0.0.0-1
ORA_RPM         = oracle-instantclient$(ORACLE_VERSION)-devel-$(RPM_VERSION).$(ARCH).rpm oracle-instantclient$(ORACLE_VERSION)-basic-$(RPM_VERSION).$(ARCH).rpm
ORA_ZIP         = instantclient-basic-linux.x64-19.5.0.0.0dbru.zip
LD_LIBRARY_PATH = /root/JUEVPROM/prometheus_oracle-master/usr/lib/oracle/$(ORACLE_VERSION)/client64/lib
BUILD_ARGS      = --build-arg VERSION=$(VERSION) --build-arg ORACLE_VERSION=$(ORACLE_VERSION)
DIST_DIR        = prometheus_oracle.$(VERSION)-ora$(ORACLE_VERSION).linux-${GOARCH}
ARCHIVE         = prometheus_oracle.$(VERSION)-ora$(ORACLE_VERSION).linux-${GOARCH}.tar.gz

#export LD_LIBRARY_PATH ORACLE_VERSION

%.rpm:
        wget -q http://yum.oracle.com/repo/OracleLinux/OL7/oracle/instantclient/$(ARCH)/getPackage/$@

%.zip:
        wget -q https://download.oracle.com/otn_software/linux/instantclient/195000/$@

download-rpms: $(ORA_RPM)

download-zips: $(ORA_ZIP)

prereq: download-rpms download-zips
        @echo deps
        sudo apt-get update
        sudo apt-get install --no-install-recommends -qq libaio1 rpm
        sudo rpm -Uvh --nodeps --force oracle*rpm
        echo $(LD_LIBRARY_PATH) | sudo tee /etc/ld.so.conf.d/oracle.conf
        sudo ldconfig

oci.pc:
        sed "s/@ORACLE_VERSION@/$(ORACLE_VERSION)/g" oci8.pc.template > oci8.pc

linux: oci.pc
        @echo build linux
        mkdir -p ./dist/$(DIST_DIR)
        PKG_CONFIG_PATH=${PWD} GOOS=linux go build $(GOFLAGS) -o ./dist/$(DIST_DIR)/prometheus_oracle
        (cd dist ; tar cfz $(ARCHIVE) $(DIST_DIR))

#darwin: oci.pc
#       @echo build darwin
#       mkdir -p ./dist/prometheus_oracle.$(VERSION).darwin-${GOARCH}
#       PKG_CONFIG_PATH=${PWD} GOOS=darwin go build $(GOFLAGS) -o ./dist/prometheus_oracle.$(VERSION).darwin-${GOARCH}/prometheus_oracle
#       (cd dist ; tar cfz prometheus_oracle.$(VERSION).darwin-${GOARCH}.tar.gz prometheus_oracle.$(VERSION).darwin-${GOARCH})

local-build:  linux

#build: docker

#deps:
#       @PKG_CONFIG_PATH=${PWD} go get

#test:
#       @echo test
#       @PKG_CONFIG_PATH=${PWD} go test $$(go list ./... | grep -v /vendor/)

#clean:
#       rm -rf ./dist sgerrand.rsa.pub glibc-2.29-r0.apk oci8.pc

#docker: ubuntu-image

#ubuntu-image: $(ORA_RPM) $(ORA_ZIP)
#       docker build $(BUILD_ARGS)  -t "juev/prometheus_oracle:$(VERSION)" .
#       docker tag "juev/prometheus_oracle:$(VERSION)" "juev/prometheus_oracle:latest"

#travis: oci.pc prereq deps test linux docker
#       @true

#.PHONY: build deps test clean docker travis oci.pc

But after this change i run make make linux 

And same errors:

sed "s/@ORACLE_VERSION@/19.5/g" oci8.pc.template > oci8.pc
build linux
mkdir -p ./dist/prometheus_oracle.0.1.1-ora19.5.linux-amd64
PKG_CONFIG_PATH=/root/JUEVPROM/prometheus_oracle-master GOOS=linux go build -ldflags "-X main.Version=0.1.1 -s -w" -o ./dist/prometheus_oracle.0.1.1-ora19.5.linux-amd64/prometheus_oracle
# github.com/mattn/go-oci8
In file included from /root/go/pkg/mod/github.com/mattn/go-oci8@v0.0.4/c_helpers.go:3:0:
./oci8.go.h:1:17: fatal error: oci.h: No such file or directory
 #include <oci.h>
                 ^
compilation terminated.
make: *** [linux] Error 2

So sad 👎

**Comment by juev** [comment]
@Seth1u I dont understand what are you doing?

Where do you get this config file? Where do you get this command line options?? And why do your run `make`??

**Comment by Seth1u** [comment]
@juev i follow this repo https://github.com/freenetdigital/prometheus_oracle_exporter. He coding an oracle exporter too

**Comment by Seth1u** [comment]
What configuration  do you use on Linux system without docker? And what metrics your exporter can print? I don’t know all metrics that you exporter can do

**Comment by juev** [comment]
@Seth1u if you used program from another programmer, ask him. Configuration file for my exporter exist in repository and comments on thi thread.

**Comment by Seth1u** [comment]
Hello @juev 
I try to compare the difference between all repositories.  And yours is most detailed.

I like it and i will continuous with yours

**Comment by Seth1u** [comment]
Hello @juev 
Thank you for all. 
I found the problem with the compilation. It was my LD_LIBRARY_PATH.
I used oracle database LD_LIBRARY_PATH for my oci8.pc.template but now i have just change it with  install of SDK and basic instantclient. 

Now i can compile exporter from your source code.

I will add some Oracle query in the main file and test it. 

Il you want i will contribute to your repo with pull request. 
I develop this exporter for Redhat Entreprise Linux and Oracle Linux.

Best regards !
